uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -116,476 +76,627 @@
+where
+    // See rust_future_new for an explanation of these trait bounds
+    F: Future<Output = T> + Send + 'static,
+    T: FfiConverter<UT> + Send + 'static,
+    UT: Send + 'static,
+{
+    fn wake(self: Arc<Self>) {
+        self.deref().wake()
     }
 
-    /// Consume a pointer to get an arc
-    ///
-    /// SAFETY: The pointer must have come from `into_raw()` or was cloned with `raw_clone()`.
-    /// Once a pointer is passed into this function, it should no longer be used.
-    fn from_raw(self_ptr: *const ()) -> Pin<Arc<Self>> {
-        unsafe { Pin::new_unchecked(Arc::from_raw(self_ptr as *const Self)) }
+    fn wake_by_ref(self: &Arc<Self>) {
+        self.deref().wake()
     }
+}
 
-    // Implement the waker interface by defining a RawWakerVTable
-    //
-    // We could also handle this by implementing the `Wake` interface, but that uses an `Arc<T>`
-    // instead of a `Pin<Arc<T>>` and in theory it could try to move the `T` value out of the arc
-    // with something like `Arc::try_unwrap()` which would break the pinning contract with
-    // `Future`.  Implementing this using `RawWakerVTable` allows us verify that this doesn't
-    // happen.
-    const RAW_WAKER_VTABLE: RawWakerVTable = RawWakerVTable::new(
-        Self::raw_clone,
-        Self::raw_wake,
-        Self::raw_wake_by_ref,
-        Self::raw_drop,
-    );
-
-    /// This function will be called when the `RawWaker` gets cloned, e.g. when
-    /// the `Waker` in which the `RawWaker` is stored gets cloned.
-    unsafe fn raw_clone(self_ptr: *const ()) -> RawWaker {
-        Arc::increment_strong_count(self_ptr as *const Self);
-        RawWaker::new(self_ptr, &Self::RAW_WAKER_VTABLE)
-    }
-
-    /// This function will be called when `wake` is called on the `Waker`. It
-    /// must wake up the task associated with this `RawWaker`.
-    unsafe fn raw_wake(self_ptr: *const ()) {
-        Self::from_raw(self_ptr).wake()
-    }
-
-    /// This function will be called when `wake_by_ref` is called on the
-    /// `Waker`. It must wake up the task associated with this `RawWaker`.
-    unsafe fn raw_wake_by_ref(self_ptr: *const ()) {
-        // This could be optimized by only incrementing the strong count if we end up calling
-        // `schedule_do_wake()`, but it's not clear that's worth the extra complexity
-        Arc::increment_strong_count(self_ptr as *const Self);
-        Self::from_raw(self_ptr).wake()
-    }
-
-    /// This function gets called when a `RawWaker` gets dropped.
-    /// This function gets called when a `RawWaker` gets dropped.
-    unsafe fn raw_drop(self_ptr: *const ()) {
-        drop(Self::from_raw(self_ptr))
+/// RustFuture FFI trait.  This allows `Arc<RustFuture<_, _, _>>` to be cast to
+/// `Arc<dyn RustFutureFFI>`, which is needed to implement the public FFI API.  In particular, this
+/// allows you to use RustFuture functionality without knowing the concrete Future type, which is
+/// unnamable.
```

Comment:
It's not really clear to me why we have this separate trait instead of storing a `dyn Future` in `RustFuture`, but that's not something I want to dig into right now. So just leaving this comment here as a note that `dyn Future` is also a thing.

Reply:
Storing `dyn Future` would be possible, but then you'd still have the other generics (`RustFuture<T, UT>`).

I think that would mean that each exported async function would need its own poll/drop/cancel/complete scaffolding functions, rather than the current situation where we have a single poll/drop/cancel, and one complete function per FFI type.  I went with this one, but it's a close call and I'm definitely open to the other way.

Reply:
I think that if each async function had its own complete function, we could drop that transmuting code, which would be nice.